### ISSUE TITLE
Request for new intensity augmentation options

### ISSUE THREAD

**Issue by sarthakpati** [fepegar/torchio#232]
**🚀 Feature**
It would be great if the following options could be added:
- [Color augmentations](https://github.com/MIC-DKFZ/batchgenerators#supported-augmentations)
  - brightness (additive, multiplivative)
  - contrast
  - gamma (like gamma correction in photo editing)

**Motivation**

These would take care of cases where the input data is not great.

**Pitch**

The aforementioned options would make DL models more robust.

**Alternatives**

Using other packages.

**Comment by romainVala** [comment]
torchio is currently working only with gray level images.

You can change the intensity dynamic, with the RescaleIntensity transform or histogram Normalization (but they are currently more for data homogenization than for data augmentation) 

I think it is a good idea to add those transform

**Comment by theabster** [comment]
@romainVala I'm not quite sure I understand. I have 3D images of proteins saved as TIFF files. The samples have all three color channels for each "slice" of the image. So each sample has dimensions of something like (40, 512, 512, 3) for a sample with 40 slices. Would torchio not work with this? Would the same transform be applied to all 3 channels implicitly, or do I need to do this myself somehow?

**Comment by romainVala** [comment]
Hello
torchio was developed with MRI data in mind, multi channel is often use to concatenate several modalities, but I have no experience with it.

From a quick look at the code, I am not sure for spatial transform @fepegar how are multiple channel handle ? `assert len(tensor) == 1` is not a restriction to single channel ?

For intensity transform, (scalling or znormalization), the statistics are compute from a single array (that contain all the channel)
would be interesting to have an option to handle channels separately 
I guess for color channel, it is a special case, and there is a need for special transforms

Note that it is quite easy to build a new transform for specific use. ... if on wants to give a try and propose a PR you are wellcome ...

**Comment by fepegar** [comment]
@sarthakpati 

> It would be great if the following options could be added:
Color augmentations
brightness (additive, multiplivative)
contrast
gamma (like gamma correction in photo editing)

As I said in #231, would you like to contribute?

The utility of some of these for medical image processing is not very clear. For example, pixel intensities in MRI are often quite arbitrary, unlike in computer vision (0-255), so there's not much point in adding (brightness) or multiplying (contrast) intensities. 

You can use the [`Lambda`](https://torchio.readthedocs.io/transforms/others.html#lambda) to easily implement custom transforms.

---

@theabster 

> So each sample has dimensions of something like (40, 512, 512, 3) for a sample with 40 slices. Would torchio not work with this? Would the same transform be applied to all 3 channels implicitly, or do I need to do this myself somehow?

Yes, it would work. For now, you need to separate the channels, but the same spatial transform will be applied to all of them.

```python
import torch
import torchio as tio
import numpy as np

stack = torch.rand(40, 512, 512, 3)
channels_first = stack.transpose(0, 3)  # put channels first
print('channels_first.shape')
print(channels_first.shape)

red, green, blue = channels_first
in_plane_spacing = 1e-3  # one micron
slice_spacing = 1  # one mm
affine = np.diag((in_plane_spacing, in_plane_spacing, slice_spacing, 1))
print('affine')
print(affine)

subject = tio.Subject(
   red=tio.ScalarImage(tensor=red, affine=affine),
   green=tio.ScalarImage(tensor=green, affine=affine),
   blue=tio.ScalarImage(tensor=blue, affine=affine),
)
transform = tio.RandomAffine()
transformed = transform(subject)  # same transform applied to all channels
```

Output:

```python
channels_first.shape
torch.Size([3, 512, 512, 40])
affine
[[0.001 0.    0.    0.   ]
 [0.    0.001 0.    0.   ]
 [0.    0.    1.    0.   ]
 [0.    0.    0.    1.   ]]
```

I have assumed that your images are RGB, that the pixel size (spacing) is 1 micron and that the slices are 1 mm apart.

**Comment by sarthakpati** [comment]
> As I said in #231, would you like to contribute?

Would love to, but unfortunately, I am hard-pressed for the next couple of months between various (and often unending) deadlines. I doubt I will be able to put something like this together in this time frame.

> The utility of some of these for medical image processing is not very clear. For example, pixel intensities in MRI are often quite arbitrary, unlike in computer vision (0-255), so there's not much point in adding (brightness) or multiplying (contrast) intensities.

You are right, for MRI. But, when looking at histopathology, these transformations would be tremendously useful!

Thanks for the specific example.

**Comment by romainVala** [comment]
for mri gamma augmentation would be interesting too (it changes the contrast) as it is used here https://arxiv.org/pdf/2003.01995.pdf (Cf eq 4)

**Comment by fepegar** [comment]
I agree a gamma augmentation could be directly applied to medical images. Here's a little test:

```python
import torchio as tio

class Gamma(tio.Lambda):
    def __init__(self, gamma):
        super().__init__(self.apply_gamma)
        self.gamma = gamma

    def apply_gamma(self, tensor):
        return tensor ** self.gamma

colin = tio.datasets.Colin27()
tio.Compose((rescale, Gamma(0.5)))(colin).t1.save('/tmp/t1_g05.nii.gz')
tio.Compose((rescale, Gamma(1)))(colin).t1.save('/tmp/t1_g1.nii.gz')
tio.Compose((rescale, Gamma(2)))(colin).t1.save('/tmp/t1_g2.nii.gz')
```

![Screenshot_1](https://user-images.githubusercontent.com/12688084/88072537-0ee53e80-cb6d-11ea-8d4a-76d2d087a4b6.png)

**Comment by romainVala** [comment]
it looks likes scalling changes ...
I think it makes a difference if the data are scaled between 0 and 1 before 
and then you should compare gamma = 0.8 and 1.2 to see different contrast change

**Comment by fepegar** [comment]
> it looks likes scaling changes ...

Not sure what you mean by scaling here.

> I think it makes a difference if the data are scaled between 0 and 1 before

I rescaled all of them before applying the gamma transform. They are displayed with a window between 0 and 1.

> and then you should compare gamma = 0.8 and 1.2 to see different contrast change

I don't follow either. Why is that more helpful than the 0.5 and 2 that I used in the example?

**Comment by fepegar** [comment]
I did forget to add the `rescale` line in the code! I have edited it.

**Comment by fepegar** [comment]
These are 0.8, 1 and 1.2:

![Screenshot_2](https://user-images.githubusercontent.com/12688084/88079672-f88fb080-cb75-11ea-9624-6a08f0459b63.png)

**Comment by romainVala** [comment]
great 
it does look like a contrast change :+1: 
0.5 2 is fine too (I just red to quick and miss the 0.5 value ...)
I think it worth a RandomTransorm, or may be we should do a RandomLambda transform ... (so that one could add any function with some parameter being random in a certain range ...
